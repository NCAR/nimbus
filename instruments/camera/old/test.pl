#!/usr/bin/perl
$i = 0;
my @array = (1,2,3,4,5,6);
splice @array, 2, 1;
splice @array, $i, 1;
print join(":", @array);
print "\n";
