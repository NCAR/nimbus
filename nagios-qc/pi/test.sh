#! / bin / bash           
output=$(psg radar)
echo "$output"
