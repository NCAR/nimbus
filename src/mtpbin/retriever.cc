#include <stdio.h>
#include <string.h>
#include <iostream>
#include "rcf_set.h"
#include "retriever.h"

using namespace std;

std::vector<float> Retriever::Retrieve(float ScanBTs[], float ACAltKm)
{
  float Temperature;
  std::vector<float> TempProf;
  RC_Set_4Retrieval BestWtdRCSet = 
                      _Rcf_Set.getBestWeightedRCSet(ScanBTs, ACAltKm, 0.0);
  for (int L = 0; L < NUM_RETR_LVLS; L++) 
  {
    Temperature = BestWtdRCSet.FL_RCs.TAvgRl[L];
    for (int j = 0; j < NUM_BRT_TEMPS; j++)
    {
      float BtDiff = (ScanBTs[j] - BestWtdRCSet.FL_RCs.MBTAvg[j]);
      Temperature = Temperature + (BestWtdRCSet.FL_RCs.RC[L][j] * BtDiff);
    }
    TempProf.push_back(Temperature);
  }

  return TempProf;
}

