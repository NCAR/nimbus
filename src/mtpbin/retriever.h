/*
 ******************************************************************************
 ** retriever.h
 **
 ** The Microwave Temperature Profiler (MTP) instrument build by JPL for
 ** NCAR as part of the set of instruments known as HAIS (build specifically
 ** in support of the Haiper GV aircraft) came with processing software 
 ** written in Visual Basic 6.0. 
 **
 ** A set of Retrieval Coefficient Files (RCFs) are currently written 
 ** by the VB program RCCalc which takes RAOB data and converts it into 
 ** Templates that describe what that RAOB profile would look like to the MTP 
 ** instrument if the instrument was used to measure the atmosphere described 
 ** by the profile. 
 **
 ** When the MTP instrument completes a scan of the atmosphere the scan counts
 ** are converted to Brightness Temperatures, using these the "best match" 
 ** RCF is found and it's corresponding Retrieval Coefficients are used
 ** to determine the atmospheric temperature profile.
 ** 
 ** This class encapsulates the retrieval method which perdforms the inverse
 ** calculation of the radiative transfer model to take a set of brightness 
 ** temperatures and generate the physical temperature profile.  It does so by
 ** using the RetrievalCoefficientFileSet class to obtain the weighted averaged
 ** retrieval coefficients from the RCF that best matches our scan at our
 ** flight altitude.
 **
 ** Copyright 2015-2016 University Corporation for Atmospheric Research
 **  - VB6 and Algorithm Copyright MJ Mahoney, NASA Jet Propulsion Laboratory
 **
*/
#ifndef _RETRIEVER_h
#define _RETRIEVER_h

#include <stdio.h>
#include <string>
#include <vector>
#include "rcf_set.h"

class Retriever
{

public:

  /**
   * Construcctor
   * @param std::string is directory name containing the RCF files to be
   * used.
   */
  Retriever(RetrievalCoefficientFileSet rcfs): _Rcf_Set(rcfs) { };

  /**
   * Retrieve function.  Obtain the Physical temperature profile from a scan.
   *
   * @param[in] is an array fo floating point values indicating the brightness
   * temperature values in Degrees Kelvin from the scan.  The first 10 values 
   * are for channel 1 at each of the scan angles (highest angle to lowest 
   * angle), the second 10 are for channel 2 and the third 10 are for channel 3.
   *
   * @param[in] is the floating point altitude of the aircraft in km
   *
   */
  std::vector<float> Retrieve(float[], float) ;

private:
  RetrievalCoefficientFileSet _Rcf_Set;

};






#endif
