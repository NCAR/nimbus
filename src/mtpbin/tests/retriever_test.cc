#include <iostream>
#include <fstream>
#include <stdlib.h>

#include "../retriever.h"
#include "test.h"

using namespace std;

/*
 * This program is designed to test the Retriever class
 * to assure that it's performing as expected.  The current output file:
 * retriever.out has been vetted to assure that it has correct information
 * resulting from the tests this program peforms.
 *
 * Usage:
 * retriever_test > retriever_test.new
 * diff retriever_test.out retriever_test.new
 *
 * There should be no differences found.  If differences are found then 
 * something has affected the functionality of the class and should be 
 * investigated and resolved.
 */

void retriever_test(float *, int, std::vector<float>, double, std::string);

main () {
  // First HIPPO test
  retriever_test(flightLevelsKmH,numFlightLevelsH,scanBTsH1,ACAltKmH1,RCFdirH);

  // Second HIPPO test
  retriever_test(flightLevelsKmH,numFlightLevelsH,scanBTsH2,ACAltKmH2,RCFdirH);

  // DEEPWAVE test
  retriever_test(flightLevelsKmD,numFlightLevelsD,scanBTsD1,ACAltKmD1,RCFdirD);

  // CSET test
  //retriever_test(flightLevelsKmC,numFlightLevelsC,scanBTsC1,ACAltKmC1,RCFdirC);
}

void retriever_test(float *flightLevelsKm, int numFlightLevels, 
	std::vector<float> scanBTs,double ACAltKm,std::string RCFdir) {

  /* Put together a functioning retrieval_coefficient_fileset */
  RetrievalCoefficientFileSet RCF_Set(RCFdir);
  RCF_Set.setFlightLevelsKm(flightLevelsKm, numFlightLevels);
  Retriever Rtr(RCF_Set);

  // Validate that the function for selecting an RCF (template) from the
  // set is still functioning properly.  Provide a couple of scan values 
  // and verify that the correct RCF is selected.
  cout<<"About to call Retrieve\n";
  AtmosphericTemperatureProfile ATP;
  ATP = Rtr.Retrieve(scanBTs, ACAltKm);

  cout<<"Profile Temperatures:\n";
  for (int i=0; i<NUM_RETR_LVLS; i++) 
  {
    cout << "[" << i << "]:" << ATP.Temperatures[i];
    if (i%5 == 0) cout << "\n";
  }
  cout<<"\n\n";

  cout<<"Profile Altitudes:\n";
  for (int i=0; i<NUM_RETR_LVLS; i++)
  {
    cout << "[" << i << "]:" << ATP.Altitudes[i];
    if (i%5 == 0) cout << "\n";
  } 
  cout<<"\n\n";

}
/******************************************************************************/

/*
  RC_Set_4Retrieval BestWtdRCSet = RCF_Set.getBestWeightedRCSet(scanBTs, 
                                                                12.49947, 0.0);
  RetrievalCoefficientFile TempRCF1(RCF_Set.getRCFbyId(BestWtdRCSet.RCFId));
  std::vector<RC_Set_1FL> FlRcSetVec;
  std::vector<int>::size_type sz;
  RC_Set_1FL AvgWtSet;

  // Provide some verification that math going into RCF selection is correct
  // by looking at the avg weighted flight level data.
  if (TempRCF1.isValid()) 
  {
    cout<<"1st BestWeightedRCSet is from file:"
        <<BestWtdRCSet.RCFFileName.c_str()<<"\n";
    FlRcSetVec = TempRCF1.getFL_RC_Vec();
    sz = FlRcSetVec.size();
    std::cout <<"PAltKm:12.49947  1st level:"<<FlRcSetVec.begin()->Palt
              <<"  last level:"<<FlRcSetVec.end()->Palt<<"  -1:"
              <<FlRcSetVec[sz-2].Palt<<"\n";
    AvgWtSet = TempRCF1.getRCAvgWt(12.49947);
    std::cout<<"Num,   Avg,  Rms\n";
    for (int i = 0; i < NUM_BRT_TEMPS; i++) 
    {
      std::cout<<i<<", "<<AvgWtSet.MBTAvg[i]<<", "<<AvgWtSet.MBTRms[i]<<"\n";
    }
    cout<<" RCwt:\n";
    for (int j = 0; j < NUM_RETR_LVLS; j++) {
      for (int i = 0; i < NUM_BRT_TEMPS; i++) {
        if (i%5 != 0) cout<<"["<<j<<"]["<<i<<"]:"<<AvgWtSet.RC[j][i];
        else cout<<"["<<j<<"]["<<i<<"]:"<<AvgWtSet.RC[j][i]<<'\n';
      }
    }
    cout <<"\n\n";
  }

  cout<<"Best Retrieval Coef for HIPPO-5 flt on 8/9/2011 at 152700 hours:\n"
      <<"  ID:"<<BestWtdRCSet.RCFId.c_str()<<"\n  "
      <<"Filename:"<<BestWtdRCSet.RCFFileName.c_str()<<"\n  "
      <<"SumLnProb:"<<BestWtdRCSet.SumLnProb<<"\n";

  // ob(1)-ob(30) values for scan at 220214 HIPPPO-5 flight date 20110901
  scanBTs   =       {262.1815, 262.0143, 262.2418, 263.3536, 264.2274,
                     264.8418, 264.9377, 265.8375, 266.8653, 268.4430,
                     263.2054, 263.5550, 263.7445, 264.3099, 264.4827, 
                     264.8418, 264.9814, 265.5672, 266.2898, 267.8371,
                     262.9489, 263.1401, 263.4613, 264.2195, 264.6548,
                     264.8418, 265.1116, 265.6754, 266.2920, 267.0385};

  cout<<"About to call getBestWeightedRCSet\n";
  BestWtdRCSet = RCF_Set.getBestWeightedRCSet(scanBTs, 2.050817, 0.0);
  RetrievalCoefficientFile TempRCF2(RCF_Set.getRCFbyId(BestWtdRCSet.RCFId));
  if (TempRCF2.isValid()) 
  {
    cout<<"2nd BestWeightedRCSet is from file:"
        <<BestWtdRCSet.RCFFileName.c_str()<<"\n";
    FlRcSetVec = TempRCF2.getFL_RC_Vec();
    sz = FlRcSetVec.size();
    std::cout <<"PAltKm:2.050817  1st level:"<<FlRcSetVec.begin()->Palt
              <<"  last level:"<<FlRcSetVec.end()->Palt<<"  -1:"
              <<FlRcSetVec[sz-2].Palt<<"\n";
    AvgWtSet = TempRCF2.getRCAvgWt(2.050817);
    std::cout<<"Num,   Avg,  Rms\n";
    for (int i = 0; i < NUM_BRT_TEMPS; i++) 
    {
      std::cout<<i<<", "<<AvgWtSet.MBTAvg[i]<<", "<<AvgWtSet.MBTRms[i]<<"\n";
    }
    cout<<" RCwt:\n";
    for (int j = 0; j < NUM_RETR_LVLS; j++) {
      for (int i = 0; i < NUM_BRT_TEMPS; i++) {
        if (i%5 != 0) cout<<"["<<j<<"]["<<i<<"]:"<<AvgWtSet.RC[j][i];
        else cout<<"["<<j<<"]["<<i<<"]:"<<AvgWtSet.RC[j][i]<<'\n';
      }
    }
    cout <<"\n\n";
  }
  

  cout<<"Best Retrieval Coef for HIPPO-5 flt on 9/1/2011 at 220214 hours:\n"
      <<"  ID:"<<BestWtdRCSet.RCFId.c_str()<<"\n  "
      <<"Filename:"<<BestWtdRCSet.RCFFileName.c_str()<<"\n  "
      <<"SumLnProb:"<<BestWtdRCSet.SumLnProb<<"\n";

*/
