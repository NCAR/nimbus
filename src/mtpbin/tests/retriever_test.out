About to call Retrieve
Profile Temperatures:
[0]:279.394
[1]:269.113[2]:258.756[3]:248.386[4]:240.695[5]:235.321
[6]:230.918[7]:227.924[8]:225.568[9]:223.776[10]:221.896
[11]:220.133[12]:219.332[13]:219.023[14]:219.181[15]:219.336
[16]:218.899[17]:217.95[18]:216.867[19]:215.915[20]:215.007
[21]:214.04[22]:213.283[23]:212.573[24]:212.011[25]:211.548
[26]:211.179[27]:211.599[28]:213.295[29]:215.982[30]:218.923
[31]:223.55[32]:229.587

